platform/android/src/style/layers/layers.cpp

```diff
@@ -25,52 +25,45 @@ namespace mbgl {
+        Layer* operator()(style::LineLayer& layer) { return new LineLayer(map, layer); }
+        Layer* operator()(style::RasterLayer& layer) { return new RasterLayer(map, layer); }
+        Layer* operator()(style::SymbolLayer& layer) { return new SymbolLayer(map, layer); }
+        Layer* operator()(style::CustomLayer& layer) { return new CustomLayer(map, layer); }
+    };
+
+    Layer* layer = coreLayer.accept(LayerPeerIntitializer {map});
+    return layer ? layer : new UnknownLayer(map, coreLayer);
 }
 
-static Layer* initializeLayerPeer(Map& map, std::unique_ptr<mbgl::style::Layer> coreLayer) {
-    if (coreLayer->is<style::BackgroundLayer>()) {
-        return createPeer<style::BackgroundLayer, BackgroundLayer>(map, std::move(coreLayer));
-    } else if (coreLayer->is<style::CircleLayer>()) {
-        return createPeer<style::CircleLayer, CircleLayer>(map, std::move(coreLayer));
-    } else if (coreLayer->is<style::FillExtrusionLayer>()) {
-        return createPeer<style::FillExtrusionLayer, FillExtrusionLayer>(map, std::move(coreLayer));
-    } else if (coreLayer->is<style::FillLayer>()) {
-        return createPeer<style::FillLayer, FillLayer>(map, std::move(coreLayer));
-    } else if (coreLayer->is<style::LineLayer>()) {
-        return createPeer<style::LineLayer, LineLayer>(map, std::move(coreLayer));
-    } else if (coreLayer->is<style::RasterLayer>()) {
-        return createPeer<style::RasterLayer, RasterLayer>(map, std::move(coreLayer));
-    } else if (coreLayer->is<style::SymbolLayer>()) {
-        return createPeer<style::SymbolLayer, SymbolLayer>(map, std::move(coreLayer));
-    } else if (coreLayer->is<mbgl::style::CustomLayer>()) {
-        return createPeer<style::SymbolLayer, SymbolLayer>(map, std::move(coreLayer));
-    } else {
-        return new UnknownLayer(map, std::move(coreLayer));
+struct UniqueLayerPeerIntitializer {
+    mbgl::Map& map;
+    std::unique_ptr<style::Layer> layer;
+
+    template <class LayerType>
+    std::unique_ptr<LayerType> cast() {
+        return std::unique_ptr<LayerType>(layer.release()->as<LayerType>());
     }
+
+    Layer* operator()(style::BackgroundLayer&) && { return new BackgroundLayer(map, cast<style::BackgroundLayer>()); }
+    Layer* operator()(style::CircleLayer&) && { return new CircleLayer(map, cast<style::CircleLayer>()); }
+    Layer* operator()(style::FillExtrusionLayer&) && { return new FillExtrusionLayer(map, cast<style::FillExtrusionLayer>()); }
+    Layer* operator()(style::FillLayer&) && { return new FillLayer(map, cast<style::FillLayer>()); }
+    Layer* operator()(style::LineLayer&) && { return new LineLayer(map, cast<style::LineLayer>()); }
+    Layer* operator()(style::RasterLayer&) && { return new RasterLayer(map, cast<style::RasterLayer>()); }
+    Layer* operator()(style::SymbolLayer&) && { return new SymbolLayer(map, cast<style::SymbolLayer>()); }
+    Layer* operator()(style::CustomLayer&) && { return new CustomLayer(map, cast<style::CustomLayer>()); }
```

Comment:
This could be done by adding a typedef member to the `mbgl::android::LineLayer` with a shared name (e.g. `using LayerType = style::LineLayer` in `mbgl::android::LineLayer`, then using a function like

```
template <class PeerType>
Layer* operator(typename PeerType::LayerType&) && {
    return new PeerType(map, cast<typename PeerType::LayerType>());
}
```